dali/operators/geometry/affine_transforms/combine_transforms.cc

```diff
@@ -0,0 +1,160 @@
+                 const workspace_t<CPUBackend> &ws) override {
+    assert(ws.NumInput() > 1);
+    TensorListShape<> in0_shape = ws.template InputRef<CPUBackend>(0).shape();
+    ndim_ = in0_shape[0][0];
+    nsamples_ = in0_shape.size();
+
+    DALI_ENFORCE(in0_shape.sample_dim() == 2 &&
+                 in0_shape.size() > 0 &&
+                 in0_shape[0][1] == (in0_shape[0][0] + 1),
+      make_string(
+        "The input, if provided, is expected to be a 2D tensor with dimensions "
+        "(ndim, ndim+1) representing an affine transform. Got: ", in0_shape));
+
+    for (int i = 0; i < ws.NumInput(); i++) {
+      const auto &shape = ws.template InputRef<CPUBackend>(i).shape();
+      DALI_ENFORCE(shape == in0_shape,
+        make_string("All input transforms are expected to have the same shape. Got: ",
+                    in0_shape, " and ", shape, "."));
+    }
+
+    input_order_.clear();
+    input_order_.resize(ws.NumInput());
+
+    // By default we visit the inputs in opposite order to the matrix multiplication order
+    // (descending indices). If we want the reverse order, we use ascending indices instead
+    std::iota(input_order_.begin(), input_order_.end(), 0);
+    if (!reverse_order_) {
+      std::reverse(input_order_.begin(), input_order_.end());  // descending order
+    }
+
+    output_descs.resize(1);  // only one output
+    output_descs[0].type = TypeTable::GetTypeInfo(dtype_);
+    output_descs[0].shape = uniform_list_shape(nsamples_, {ndim_, ndim_+1});
+    return true;
+  }
+
+  template <typename T>
+  void RunImplTyped(workspace_t<CPUBackend> &ws, dims<>) {
+    DALI_FAIL(make_string("Unsupported number of dimensions ", ndim_));
+  }
+
+  template <typename T, int ndim, int... ndims>
+  void RunImplTyped(workspace_t<CPUBackend> &ws, dims<ndim, ndims...>) {
+    if (ndim_ != ndim) {
+      RunImplTyped<T>(ws, dims<ndims...>());
+      return;
+    }
+
+    constexpr int mat_dim = ndim + 1;
+    auto &out = ws.template OutputRef<CPUBackend>(0);
+    out.SetLayout({});  // no layout
+    auto out_view = view<T>(out);
+
+    matrix_data_.Resize({nsamples_ * static_cast<int64_t>(sizeof(affine_mat_t<T, mat_dim>))});
+    span<affine_mat_t<T, mat_dim>> matrices{
+        reinterpret_cast<affine_mat_t<T, mat_dim> *>(matrix_data_.mutable_data<T>()), nsamples_};
+    for (auto &mat : matrices) {
+      mat = affine_mat_t<T, mat_dim>::identity();
+    }
+
+    for (auto input_idx : input_order_) {
+      auto &in = ws.template InputRef<CPUBackend>(input_idx);
+      auto in_view = view<T>(in);
+      for (int sample_idx = 0; sample_idx < nsamples_; sample_idx++) {
+        auto &mat = matrices[sample_idx];
+        auto next_mat = affine_mat_t<T, mat_dim>::identity();
+        for (int i = 0, k = 0; i < ndim; i++) {
+          for (int j = 0; j < ndim + 1; j++, k++) {
+            next_mat(i, j) = in_view[sample_idx].data[k];
+          }
+        }
+        mat = next_mat * mat;  // mat mul
```

Comment:
```suggestion
        mat = reverse_ ? mat * next_mat : next_mat * mat;  // mat mul
```
This obviates the need for input_order_ and the condition should not pose too much overhead. With reversed loop nesting order, `mat` could be a local variable that's stored immediately in the output (no need for `matrices`).